sql/core/src/main/scala/org/apache/spark/sql/execution/ProjectionOverSchema.scala

```diff
@@ -38,7 +38,7 @@ private[execution] case class ProjectionOverSchema(schema: StructType) {
       case GetArrayItem(child, arrayItemOrdinal) =>
         getProjection(child).map { projection => GetArrayItem(projection, arrayItemOrdinal) }
       case a: GetArrayStructFields =>
-        getProjection(a.child).map(p => (p, p.dataType)).map {
+        getProjection(a.child).map(p => (p, p.dataType)).collect {
```

Comment:
How about this? IMO `.collect` can't catch illegal inputs?
```
        getProjection(a.child).map(p => (p, p.dataType)).map {
          case (projection, ArrayType(projSchema @ StructType(_), _)) =>
            GetArrayStructFields(projection,
              projSchema(a.field.name),
              projSchema.fieldIndex(a.field.name),
              projSchema.size,
              a.containsNull)
          case _ =>
            sys.error("....")
        }
```

Reply:
Well, there is a semantic change using `collect`. I will provide a unit test to verify the change or simply using `sys.error`

Reply:
Does this generate an error or just a warning in 2.12? It's an unmatched case error, right? 
We had 'fixed' this elsewhere with `case _ => throw new IllegalStateException(...)` or something. At runtime it would have caused a MatchError before, right? Because it didn't, I hope, or else tests would have failed, it matters a bit less what happens in this case. But it felt more conservative to preserve the error behavior; maybe I was wrong about that?

Reply:
To use `collect`, one must dive into `SPARK-4502`.

This PR is intended for #22264 .

I will preserve the error behavior.